### ISSUE TITLE
SQL stats not collected on Payara 4

### ISSUE THREAD

**Issue by tomekkup** [javamelody/javamelody#1141]
JavaMelody seems to not collect JDBC and SQL stats on Payara 4. I've configured filter, listener, reconfigured jdbc pools to use ObjectFactory. Already put javamelody, objectfactory and jrobin jars into lib folder and still it shows empty stats.

Any idea ???

```xml
<filter>
        <filter-name>Java Melody Filter</filter-name>
        <filter-class>net.bull.javamelody.MonitoringFilter</filter-class>
        <async-supported>true</async-supported>
        <init-param>
            <param-name>url-exclude-pattern</param-name>
            <param-value>/static/.*</param-value>
        </init-param>
        <init-param>
            <param-name>displayed-counters</param-name>
            <param-value>http,jsf,ejb,sql,error,log</param-value>
        </init-param>
        <init-param>
            <param-name>datasources</param-name>
            <param-value>jdbc/CouncilDS</param-value>
        </init-param>
        <init-param>
            <param-name>authorized-users</param-name>
            <param-value>admin:{SHA-256}************</param-value>
        </init-param>
</filter>

<filter-mapping>
        <filter-name>Java Melody Filter</filter-name>
        <url-pattern>/*</url-pattern>
        <dispatcher>REQUEST</dispatcher>
        <dispatcher>ASYNC</dispatcher>
</filter-mapping>

<listener>
    <listener-class>net.bull.javamelody.SessionListener</listener-class>
</listener>
```

![obraz](https://user-images.githubusercontent.com/1681135/215999530-30b80125-5925-40e2-a616-3ef82eda641b.png)

```09:39:09.478 [admin-thread-pool::admin-listener(1)] INFO  f.p.micro.cdi.extension.ClusteredCDIEventBusImpl - Clustered CDI Event bus initialized for 
09:39:09.478 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - JavaMelody listener init started
09:39:09.487 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - datasources found in JNDI: [jdbc/__TimerPool__pm, jdbc/ActivitiDS, jdbc/ActivitiDS_uncached, jdbc/CouncilDS_uncached, jdbc/ActivitiDS_uncached__pm, jdbc/CouncilDS_uncached__pm, jdbc/__TimerPool, jdbc/CouncilDS]
09:39:09.488 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource needs rewrap: jdbc/__TimerPool__pm of class com.sun.gjc.spi.jdbc40.DataSource40
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource rewrapped: jdbc/__TimerPool__pm
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource needs rewrap: jdbc/ActivitiDS of class com.sun.gjc.spi.jdbc40.DataSource40
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource rewrapped: jdbc/ActivitiDS
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource needs rewrap: jdbc/ActivitiDS_uncached of class com.sun.gjc.spi.jdbc40.DataSource40
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource rewrapped: jdbc/ActivitiDS_uncached
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource needs rewrap: jdbc/CouncilDS_uncached of class com.sun.gjc.spi.jdbc40.DataSource40
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource rewrapped: jdbc/CouncilDS_uncached
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource needs rewrap: jdbc/ActivitiDS_uncached__pm of class com.sun.gjc.spi.jdbc40.DataSource40
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource rewrapped: jdbc/ActivitiDS_uncached__pm
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource needs rewrap: jdbc/CouncilDS_uncached__pm of class com.sun.gjc.spi.jdbc40.DataSource40
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource rewrapped: jdbc/CouncilDS_uncached__pm
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource needs rewrap: jdbc/__TimerPool of class com.sun.gjc.spi.jdbc40.DataSource40
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource rewrapped: jdbc/__TimerPool
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource needs rewrap: jdbc/CouncilDS of class com.sun.gjc.spi.jdbc40.DataSource40
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - Datasource rewrapped: jdbc/CouncilDS
09:39:09.491 [admin-thread-pool::admin-listener(1)] DEBUG net.bull.javamelody - JavaMelody listener init done in 13 ms```

**Comment by evernat** [comment]
Since you said you are using ObjectFactory, I suppose that you have done exactly like https://github.com/javamelody/javamelody/wiki/UserGuideAdvanced#monitoring-of-sql-requests-and-of-jdbc-connections-in-glassfish-v3
And your log above seems to confirm that.
You are supposed to put javamelody and jrobin jar files in your webapp, not into lib folder (but I suppose that it will not change the problem of this issue).

Glassfish datasources are known to me to be of class `com.sun.gjc.spi.jdbc40.DataSource40`. And that is the relevant code in javamelody which supports `com.sun.gjc.spi.jdbc40.DataSource40` : https://github.com/javamelody/javamelody/blob/master/javamelody-core/src/main/java/net/bull/javamelody/JdbcWrapper.java#L499

Can you say me what is the class of a Payara 4 datasource ?

**Comment by evernat** [comment]
I mean the class after lookup in JNDI probably like this : `new InitialContext().lookup("java:jdbc/ActivitiDS").getClass().getName()`

**Comment by tomekkup** [comment]
at first I got `javax.naming.NameNotFoundException: No object bound to name java:jdbc/ActivitiDS` . 

Correct jndi name is just `jdbc/ActivitiDS` and className is `com.sun.gjc.spi.jdbc40.DataSource40`

**Comment by evernat** [comment]
It is supposed to work. That is because I suppose that its serverInfo contains `Payara` and like Glassfish, a datasource is of type `com.sun.gjc.spi.jdbc40.DataSource40` which is supported and it contains in its superclass a private field named `cm` which is supported. Logs above shows no issue with all that.
And a connection is probably of class `com.sun.gjc.spi.jdbc40.ConnectionHolder40` or `com.sun.gjc.spi.jdbc40.ConnectionWrapper40`.

But when you get a connection from the datasource, can you confirm that it is of class `com.sun.gjc.spi.jdbc40.ConnectionHolder40` or `com.sun.gjc.spi.jdbc40.ConnectionWrapper40` ? Or is it of class `com.sun.gjc.spi.jdbc40.ProfiledConnectionWrapper40` or something else ?

If the latter, the rewrapConnection method in the JdbcWrapper class of javamelody should be adapted.

Thanks.

**Comment by evernat** [comment]
And do you have `wrap-jdbc-objects` set to true in your connection pool configuration?

**Comment by tomekkup** [comment]
Class name is exactly `com.sun.gjc.spi.jdbc40.ProfiledConnectionWrapper40`

**Comment by tomekkup** [comment]
> And do you have `wrap-jdbc-objects` set to true in your connection pool configuration and request tracing or slow query enabled in Payara ?

yes

**Comment by evernat** [comment]
fixed by ae943e5 in 1.x branch and by 50640bc in 2.x branch.

A snapshot build is available at
http://javamelody.org/downloads/javamelody-core-1.95.0-SNAPSHOT.jar
